### ISSUE TITLE
Change number of worker threads from within runtime

### ISSUE THREAD

**Issue by SergioBenitez** [tokio-rs/tokio#3329]
In Rocket, the desired number of worker threads is dynamically configurable. Below is a complete Rocket application that sets the number of workers to the result of some async function `some_fn` and then starts the server (`rocket::main` can be thought of as `tokio::main`):

```rust
#[rocket::main]
async fn main() {
    let figment = Config::figment()
        .merge(("workers", some_fn().await));

    rocket::custom(figment)
        .mount("/", routes![/* .. */])
        .launch()
        .await
}
```

The hope is that Rocket can set the number of worker threads used by the async executor to the `workers` value. At present, the number of worker threads can only be configured before the runtime is started. But Rocket only learns the `workers` value _after_ the `main` `Future` is executed, which of course requires an already running runtime, at which point it is too late to set the value. Thus, in order to support this functionality, `tokio` would need to provide a mechanism to dynamically configure the number of worker threads from an already running runtime.

I'll note that I considered a solution of the following form, where the signature for `main` is required to return a `Rocket`:

```rust
#[rocket::launch]
async fn main() -> Rocket {
    let figment = Config::figment()
        .merge(("workers", some_fn().await));

    rocket::custom(figment)
        .mount("/", routes![/* .. */])
}
```

The idea here is that the wrapper code generated by `launch` would read the `workers` config value from the `Rocket` instance returned by `main()` , set the number of worker threads in the executor accordingly, and then run the `launch()` future in the appropriately configured executor. However, the issue here is the same: we need an executor to run the `main` future, and that executor _must_ be the same as the executor in which `launch()` is run due to potential state generated by other tasks in the `main` future. As such, we continue to require `tokio` to allow dynamic configuration of worker values.

**Comment by Darksonn** [comment]
I suppose this could be added just as a method on the `Runtime` object.

**Comment by carllerche** [comment]
Changing the number of workers at runtime would not be a trivial change (3/5 difficulty). Much of the scheduler's algorithm assumes the number will not change at runtime. The strategy would have to be to shutdown the executor and start a new one with a different number of threads.

One thought: Could you use an executor to bootstrap rocket configuration and transition to a tokio executor once you have the configuration?

**Comment by SergioBenitez** [comment]
> One thought: Could you use an executor to bootstrap rocket configuration and transition to a tokio executor once you have the configuration?

My understanding, and please do correct if this is wrong, is that certain state, like timers, is runtime-instance sensitive. Thus, the state is the "bootstrap executor" will not be available in the "tokio executor" as required by certain tasks such as database pooling.

**Comment by carllerche** [comment]
You are correct. We could probably have a way to "restart" the runtime w/ a new configuration while maintaining the reactors in order to ensure continuity. It would require shutting down any existing tasks.

**Comment by SergioBenitez** [comment]
That'd be great. Or even just a way to create a new runtime that inherits/moves reactors from an existing runtime, without necessarily keeping the old runtime around.
